packages/driver/src/cy/assertions.js

```diff
@@ -0,0 +1,509 @@
+      }
+
+      // reset the assertion index back to 0
+      // so we can track assertions and merge
+      // them up with existing ones
+      cmd.set('assertionIndex', 0)
+
+      return cmd.get('fn').originalFn.apply(
+        state('ctx'),
+        [subject].concat(cmd.get('args'))
+      )
+    })
+  }
+
+  const finishAssertions = (assertions) => {
+    return _.each(assertions, (log) => {
+      log.snapshot()
+
+      const e = log.get('_error')
+
+      if (e) {
+        return log.error(e)
+      }
+
+      return log.end()
+    })
+  }
+
+  const verifyUpcomingAssertions = function (subject, options = {}, callbacks = {}) {
+    const cmds = getUpcomingAssertions()
+
+    state('upcomingAssertions', cmds)
+
+    // we're applying the default assertion in the
+    // case where there are no upcoming assertion commands
+    const isDefaultAssertionErr = cmds.length === 0
+
+    if (options.assertions == null) {
+      options.assertions = []
+    }
+
+    _.defaults(callbacks, {
+      ensureExistenceFor: 'dom',
+    })
+
+    const ensureExistence = function () {
+      // by default, ensure existence for dom subjects,
+      // but not non-dom subjects
+      switch (callbacks.ensureExistenceFor) {
+        case 'dom': {
+          const $el = determineEl(options.$el, subject)
+
+          if (!$dom.isJquery($el)) {
+            return
+          }
+
+          return cy.ensureElExistence($el)
+        }
+        case 'subject':
+          return cy.ensureExistence(subject)
+
+        default:
+          return
+      }
+    }
+
+    const determineEl = function ($el, subject) {
+      // prefer $el unless it is strickly undefined
+      if (!_.isUndefined($el)) {
+        return $el
+      }
+
+      return subject
+    }
+
+    const onPassFn = () => {
+      if (_.isFunction(callbacks.onPass)) {
+        return callbacks.onPass.call(this, cmds, options.assertions)
+      }
+
+      return subject
+    }
+
+    const onFailFn = (err) => {
+      // when we fail for whatever reason we need to
+      // check to see if we would firstly fail if
+      // we don't have an el in existence. what this
+      // catches are assertions downstream about an
+      // elements existence but the element never
+      // exists in the first place. this will probably
+      // ensure the error is about existence not about
+      // the downstream assertion.
+      try {
+        ensureExistence()
+      } catch (e2) {
+        err = e2
+      }
+
+      options.error = err
+
+      if (err.retry === false) {
+        throw err
+      }
+
+      const { onFail, onRetry } = callbacks
+
+      if (!onFail && !onRetry) {
+        throw err
+      }
+
+      // if our onFail throws then capture it
+      // and finish the assertions and then throw
+      // it again
+      try {
+        if (_.isFunction(onFail)) {
+          // pass in the err and the upcoming assertion commands
+          onFail.call(this, err, isDefaultAssertionErr, cmds)
```

Comment:
Is this calling the correct `this` with the `=>`?

Reply:
In CoffeeScript code, it was an arrow function like this:

```coffee
onFailFn = (err) =>
  ## when we fail for whatever reason we need to
  ## check to see if we would firstly fail if
```